Bug fix: Vertical images in experimental lightbox

artemiomorales opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
It revises the CSS to fix the display of vertical images in the [experimental lightbox](https://github.com/WordPress/gutenberg/pull/50373).

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Address #51125 
Vertical images should be contained within the viewport when being viewed in the lightbox.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
It revises the CSS to make sure the image doesn't expand beyond the appropriate borders of the lightbox. It also improves the style for the caption.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
1. Ensure the Interactivity API and Behaviors UI setting is enabled in Gutenberg > Experiments in the admin.
2. Create a new post and an image whose height is larger than its width; set the image's resolution to _large_ or _full size_.
3. Activate the image's lightbox via Advanced > Behaviors > Lightbox.
4. Publish and view the post.
5. Click on the image; see that the image is fully visible and contained within the lightbox.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A

## Screenshots or screencast <!-- if applicable -->

artemiomorales committed:
Revise CSS to contain images within overlay properly

SantosGuillamot commented on packages/block-library/src/image/style.scss:
```diff
@@ -195,14 +195,28 @@
+			max-width: 100%;
+			max-height: 100%;
+			width: auto;
+			min-height: 0;
+			min-width: 0;
```
Do we need this? Is not 0 the default value?

artemiomorales replied:
Yes, we need this, as the default value is actually `auto`. Without it, the image will expand take up the minimum size of its content and overflow into its parent container, breaking the flexbox layout.

SantosGuillamot replied:
Oh, I read [here](https://www.w3schools.com/cssref/pr_dim_min-height.php#:~:text=Show%20demo%20%E2%9D%AF-,Default%20value%3A,-0) that the default value is true, and removing those lines didn't seem to affect. That's why I thought they weren't necessary.

artemiomorales replied:
Interesting — [MDN says](https://developer.mozilla.org/en-US/docs/Web/CSS/min-height) that the initial value is `auto`, and the behavior I saw was that of `auto` as far as I can tell.

SantosGuillamot reviewed (approved):
Approving with a small comment. I've tested it and it is definitely an improvement.

I have one question: right now, if you have a really big image with the lightbox (for example if you set it to full width or align wide), it gets smaller and it doesn't cover as much space as possible when you click on it. Is that the expected behavior or should it get the max-height?

artemiomorales committed:
Set horizontal images to occupy space of container

artemiomorales committed:
Make square images fit by height rather than expand to full width

artemiomorales commented:
> I have one question: right now, if you have a really big image with the lightbox (for example if you set it to full width or align wide), it gets smaller and it doesn't cover as much space as possible when you click on it. Is that the expected behavior or should it get the max-height?

@SantosGuillamot I believe it depended on the size of the image — if it was wide enough to cover the full screen, then it would have done so and been capped at the size of the container.

That being said, based on your comment, I decided to add another adjustment so that horizontal images expand to fit the full width of the container. This will stretch and distort thumbnails since we're not yet putting the full resolution image in the lightbox ([related issue here](https://github.com/WordPress/gutenberg/issues/51171)), but is maybe more in line with expected behavior of bigger horizontal images for now.

Do you think this is a good change, or should we revert to the way it was working before?

SantosGuillamot commented:
I will review it in more detail, but I'd like to clarify first what is the expected behavior of the lightbox.

Should it always display the largest available size of image as explained [here](https://github.com/WordPress/gutenberg/issues/51171)? My guess, from a total lack of knowledge, is that it should. We shouldn't make the image bigger than expected I assume, which is currently happening with this implementation. I may be mistaken but I believe it should behave as the attachment page but without going to a new link. Is that correct?

SantosGuillamot commented:
I was thinking a bit more about this and I believe this can be fixed by removing the sizes property from the overlay image. This way, it should get the highest size possible and it will fit into the page.

What about going back to the previous implementation and doing something like this:

```php
// Add directive to remove the sizes limitation.
$m = new WP_HTML_Tag_Processor( $content );
$m->next_tag( 'img' );
$m->set_attribute( 'sizes', '' );
$modal_content = $m->get_updated_html();
```

I quickly tested and it seems to work fine. I would probably remove the padding we are adding.

The main problem with this approach is that it would download the biggest image possible, so maybe it is not right. It would be great if we download the lightbox image when it is clicked/hovered.

Additionally, while comparing it with the media pages I wondered, should the lightbox include the caption? 🤔

SantosGuillamot commented:
Sorry for the spam 😄 I've been thinking a bit more.

> The main problem with this approach is that it would download the biggest image possible, so maybe it is not right. It would be great if we download the lightbox image when it is clicked/hovered.

This could be solved using directives that only change the attribute when the lightbox is initialized. At this moment, it could be done this way:

_In PHP_

```php
$m = new WP_HTML_Tag_Processor( $content );
$m->next_tag( 'img' );
$m->set_attribute( 'data-wp-effect', 'effects.core.image.initLightboxImage' );
$modal_content = $m->get_updated_html();
```

And the `initLightboxImage` could be like this:

```js
initLightboxImage: ( { context, ref } ) => {
	if ( context.core.image.initialized ) {
		ref.setAttribute( 'sizes', '' );
	}
},
```

---
Ideally, I believe we should use a selector that reads the current `ref`, but that isn't possible right now. We could do something more declarative like this:

_In PHP_
```php
$m = new WP_HTML_Tag_Processor( $content );
$m->next_tag( 'img' );
$m->set_attribute( 'data-wp-bind--sizes', 'selectors.core.image.imageSizes' );
$modal_content = $m->get_updated_html();
```

And the selectors could be like this:

```js
imagesSizes: ( { context, ref } ) => {
    if ( context.core.image.initialized ) {
        return '';
    } else {
        return ref.sizes;
    }
}					
```

artemiomorales commented:
@SantosGuillamot Thanks for the ideas! The issue I see with this approach is that it doesn't fully solve the issue, as the `src` and `srcset` attributes vary depending on which resolution is selected. Here's the markup for the same image with full size, large, medium, and thumbnail resolutions:

<img width="1063" alt="Screen Shot 2023-06-02 at 10 14 44 AM" src="https://github.com/WordPress/gutenberg/assets/5360536/8a58dc8d-ec88-4590-9b62-a433c8ca5461">

SantosGuillamot commented:
> The issue I see with this approach is that it doesn't fully solve the issue, as the src and srcset attributes vary depending on which resolution is selected.

You're right the values may differ depending on the settings.

Going back to my previous question: What is the purpose of the lightbox? How is this supposed to work with the lightbox? With the current implementation, if I am not mistaken, if I add a low resolution image it creates a big pixelated image if the width is higher than the height.

I have other ideas to make it work like the image media page but, is that what we want? Do we just want to show the same image resolution but with an overlay? Or is it okay pixelating an image?

---
EDIT: Sharing here an idea that could work like the media pages, but I'd like to answer the questions before talking about it and exploring deeper the potential solutions:

Adding the attachment URL to the context in PHP and change the lightbox src depending on it. Adding a selector to check if it has been initialized and don't load the image until then.

```php
// Add directive to expand modal image if appropriate.
$m = new WP_HTML_Tag_Processor( $content );
$m->next_tag( 'img' );
$m->set_attribute( 'data-wp-context', '{ "core": { "image": { "imageSrc": "' . wp_get_attachment_url($attributes['id']) . '"} } }' );
$m->set_attribute( 'data-wp-bind--src', 'selectors.core.image.imageSrc' );
$modal_content = $m->get_updated_html();
```

And in JS:

```js
imageSrc: ( { context } ) => {
	return context.core.image.initialized
		? context.core.image.imageSrc
		: '';
},
```

artemiomorales commented:
@SantosGuillamot My expectation when I open a lightbox is that it should display the highest resolution version of the image possible appropriate for my screen

If the image is a thumbnail in the content, I believe the lightbox shouldn't display a thumbnail-sized image, but should instead show an image that occupies the full width of the screen at high resolution.

I don't believe it's alright to pixelate the image; what I pushed up was meant to be a point of discussion and maybe a stopgap before fixing the issue for real.

artemiomorales commented:
> Additionally, while comparing it with the media pages I wondered, should the lightbox include the caption? 🤔

@SantosGuillamot If I recall correctly, there may be accessibility reasons for having it there — here I'm pinging in @andrewhayward for insight

artemiomorales committed:
Revise so lightbox loads original image; remove padding and hide caption

The lightbox will now load the original image when opened;
the src attribute is set dynamically so it will not be
unnecessarily loaded by the browser.

In addition, removed the padding and hid the caption.

artemiomorales commented:
@SantosGuillamot and I spoke this morning and have an implementation that we think will fix this issue as well as address #51171 and #51170.

When opened, the lightbox now loads the image's original attachment URL as the `src`, which means it gets the largest available file, addressing #51171. Importantly, this `src` is set dynamically so the image is not loaded unnecessarily by the browser, and it will not stretch the image either.

Removing the caption and padding allows the image to be displayed as large as possible, addressing #51170. 

In effect, with these changes, the lightbox acts similarly to the media page, which seems to be a good guideline for expected behavior from the lightbox.

Pinging @jasmussen @SaxonF @jameskoster for any additional feedback.

artemiomorales committed:
Fix PHP spacing for linter

artemiomorales committed:
Fix test; image src is now only set after lightbox opens

artemiomorales commented:
With the new [close button design](https://github.com/WordPress/gutenberg/issues/51126#issuecomment-1573835454), I think it makes sense to move these changes to https://github.com/WordPress/gutenberg/pull/51206 and continue the discussion regarding all these design fixes there at the same time. Just let me know if you have any questions!
